tests/src/test/scala/com/nvidia/spark/rapids/RapidsBufferCatalogSuite.scala

```diff
@@ -320,6 +324,125 @@ class RapidsBufferCatalogSuite extends AnyFunSuite with MockitoSugar {
     verify(buffer3).free()
   }
 
+  test("concurrently acquiring buffer and registering new buffer with the same ID") {
+    val catalog = new RapidsBufferCatalog
+    val bufferId = MockBufferId(5)
+    val hostBuffer = mockBuffer(bufferId, tier = HOST)
+    val deviceBuffer = mockBuffer(bufferId, tier = DEVICE)
+    catalog.registerNewBuffer(hostBuffer)
+    val handle = catalog.makeNewHandle(bufferId, -1)
+
+    val ec = ExecutionContext.global
+    val latch = new CountDownLatch(1)
+
+    val future: Future[_] = Future {
+      latch.countDown()
+      catalog.registerNewBuffer(deviceBuffer)
+    }(ec)
+
+    latch.await()
+    // Let's wait for a little more, so that registerNewBuffer is likely executed first.
+    Thread.sleep(300)
+    val acquired = catalog.acquireBuffer(handle)
+    Await.ready(future, 1.second)
+    assertResult(StorageTier.DEVICE)(acquired.storageTier)
+  }
+
+  test("concurrently setting spill priority and acquiring buffer") {
+    val catalog = new RapidsBufferCatalog
+    val bufferId = MockBufferId(5)
+    val deviceBuffer = mockBuffer(bufferId, tier = DEVICE)
+    catalog.registerNewBuffer(deviceBuffer)
+    val handle = catalog.makeNewHandle(bufferId, -1)
+
+    val ec = ExecutionContext.global
+    val latch = new CountDownLatch(1)
+    val priority: Long = 54321
+
+    val future: Future[_] = Future {
+      latch.countDown()
+      deviceBuffer.setSpillPriority(priority) // some random priority
+    }(ec)
+
+    latch.await()
+    // Let's wait for a little more, so that setSpillPriority is likely executed first.
+    Thread.sleep(300)
+    val acquired = catalog.acquireBuffer(handle)
+    Await.ready(future, 1.second)
+    assertResult(priority)(acquired.getSpillPriority)
+  }
+
+  test("currently acquiring buffer with tier and removing it") {
+    val catalog = new RapidsBufferCatalog
+    val bufferId = MockBufferId(5)
+    val deviceBuffer = mockBuffer(bufferId, tier = DEVICE)
+    catalog.registerNewBuffer(deviceBuffer)
+    val handle = catalog.makeNewHandle(bufferId, -1)
+
+    val ec = ExecutionContext.global
+    val latch = new CountDownLatch(1)
+
+    val future: Future[_] = Future {
+      latch.countDown()
+      catalog.removeBufferTier(bufferId, StorageTier.DEVICE)
+    }(ec)
+
+    latch.await()
+    // Let's wait for a little more, so that removeBufferTier is likely executed first.
+    Thread.sleep(300)
+    val caught = intercept[NoSuchElementException]{
+      catalog.acquireBuffer(handle)
+    }
+    Await.ready(future, 1.second)
+    assert(caught.getMessage == s"Cannot locate buffers associated with ID: ${bufferId}")
+  }
+
+  test("currently updating tier and getting spill state") {
```

Comment:
`updateTiers` is called from `synchronousSpill` which is running under the catalog lock. 
`isBufferSpilled` is called from `synchronousSpill` as well up the stack.

`updateTiers` is also called from `unspillBufferToHostStore` but that takes the catalog lock as well.

Reply:
As for `isBufferSpilled`, if it should be called only from there, it should be defined as private. I can make this change in this PR.

Reply:
+1 absolutely agree.